### PR TITLE
.Net: Function choice behavior in SK YAML prompts

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
### Motivation and Context
Today, function choice behavior can be configured via code and in SK JSON prompts. However, SK YAML prompts are not supported yet. This PR adds support for function choice behavior in YAML prompts.

### Description
1. The `PromptExecutionSettingsTypeConverter` replaces the `PromptExecutionSettingsNodeDeserializer` to configure polymorphic deserialization of `FunctionChoiceBehavior`.  
2. The `PromptExecutionSettingsTypeConverter` uses reflection to extract the discriminator and corresponding data type to avoid marking those classes as public and keep them internal instead. This should give us the freedom and flexibility to change them if/when necessary.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.OpenAI.UnitTests/Settings/OpenAIPromptExecutionSettingsTests.cs [hunk 1]
```diff
@@ -253,6 +253,22 @@ public void FromExecutionSettingsWithDataDoesNotIncludeEmptyStopSequences()
...
     }
 
+    [Fact]
+    public void ItRestoresOriginalFunctionChoiceBehavior()
+    {
+        // Arrange
+        var functionChoiceBehavior = FunctionChoiceBehavior.None();
+
+        var originalExecutionSettings = new PromptExecutionSettings();
+        originalExecutionSettings.FunctionChoiceBehavior = functionChoiceBehavior;
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.OpenAI.UnitTests/Settings/OpenAIPromptExecutionSettingsTests.cs:263]: nit
```suggestion
        PromptExecutionSettings originalExecutionSettings = new() { FunctionChoiceBehavior = functionChoiceBehavior };
```

File: dotnet/src/Functions/Functions.UnitTests/Yaml/Functions/KernelFunctionYamlTests.cs [hunk 2]
```diff
@@ -82,6 +83,81 @@ public void ItShouldSupportCreatingOpenAIExecutionSettings()
         Assert.Equal(0.0, executionSettings.TopP);
     }
 
+    [Fact]
+    public void ItShouldDeserializeFunctionChoiceBehaviors()
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.UnitTests/Yaml/Functions/KernelFunctionYamlTests.cs:87]: Suggest splitting this Test into different choices promoting to a Theory or individual Facts.

```suggestion
    public void ItShouldDeserializeFunctionChoiceBehaviors()
```

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Allows custom deserialization for <see cref="PromptExecutionSettings"/> from YAML prompts.
+/// </summary>
+internal sealed class PromptExecutionSettingsTypeConverter : IYamlTypeConverter
+{
+    /// <inheritdoc/>
+    public bool Accepts(Type type)
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:21]: Good case for `static`

```suggestion
    public static bool Accepts(Type type)
```

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        return type == typeof(PromptExecutionSettings);
+    }
+
+    /// <inheritdoc/>
+    public object? ReadYaml(IParser parser, Type type)
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:27]: Wonder if they have `async` methods for parsing, and using those instead on synchronous ones.

```suggestion
    public object? ReadYaml(IParser parser, Type type)
```

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        while (parser.Current is not MappingEnd)
+        {
+            var propertyName = parser.Consume<Scalar>().Value;
+            switch (propertyName)
+            {
+                case "model_id":
+                    executionSettings.ModelId = s_deserializer.Deserialize<string>(parser);
+                    break;
+                case "function_choice_behavior":
+#pragma warning disable SKEXP0001
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:47]: Suggest adding this exception in the `csproj` level and removing all instances of `pragmas` for `SKEX00NN`.

```suggestion
#pragma warning disable SKEXP0001
```

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            if (string.IsNullOrEmpty(discriminator))
+            {
+                throw new InvalidOperationException($"Type discriminator is not specified for the {derivedTypeAttribute.DerivedType} type.");
+            }
+
+            discriminatorTypeMapping.Add(discriminator!, derivedTypeAttribute.DerivedType);
+        }
+
+        options.AddKeyValueTypeDiscriminator<FunctionChoiceBehavior>(discriminatorKey!, discriminatorTypeMapping);
+#pragma warning restore SKEXP0010
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:98]: Different closing pragma (recommend removing all of them) as previous comment.
```suggestion
#pragma warning restore SKEXP0010
```

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            }
+        }
+        parser.MoveNext(); // Move past the MappingEnd event  
+        return executionSettings;
+    }
+
+    /// <inheritdoc/>
+    public void WriteYaml(IEmitter emitter, object? value, Type type)
+    {
+        throw new NotImplementedException();
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:63]: Do we plan in Serializing back to yml? Future PR?

```suggestion
        throw new NotImplementedException();
```

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs (hunk 6 above)
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:98]: Remove pragmas 😊
```suggestion
#pragma warning restore SKEXP0001
```

File: dotnet/src/Functions/Functions.UnitTests/Yaml/PromptExecutionSettingsTypeConverterTests.cs [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+namespace SemanticKernel.Functions.UnitTests.Yaml;
+
+/// <summary>
+/// Tests for <see cref="PromptExecutionSettingsTypeConverter"/>.
+/// </summary>
+public sealed class PromptExecutionSettingsTypeConverterTests
+{
+    private readonly IDeserializer _deserializer;
+
+    private readonly Kernel _kernel;
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.UnitTests/Yaml/PromptExecutionSettingsTypeConverterTests.cs:17]: nit. (Im very used to see private fields in the beggining)

Some privates methods in the end, some fields in the beggining, suggest moving those to the end

File: dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAIChatCompletion_NoneFunctionChoiceBehaviorTests.cs [hunk 9]
```diff
@@ -52,40 +52,40 @@ public async Task SpecifiedInCodeInstructsConnectorNotToInvokeKernelFunctionAsyn
...
-    //        """";
+        var promptTemplate = """"
+            template_format: semantic-kernel
+            template: How many days until Christmas?
+            execution_settings:
+              default:
+                temperature: 0.1
+                function_choice_behavior:
+                  type: none
+            """";
```
Comment (RogerBarreto) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAIChatCompletion_NoneFunctionChoiceBehaviorTests.cs:77]: nit.

```suggestion
        var promptTemplate = """
            template_format: semantic-kernel
            template: How many days until Christmas?
            execution_settings:
              default:
                temperature: 0.1
                function_choice_behavior:
                  type: none
            """;
```

File: dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAIChatCompletion_AutoFunctionChoiceBehaviorTests.cs [hunk 10]
```diff
@@ -165,46 +165,46 @@ public async Task SpecifiedInCodeInstructsConnectorToInvokeKernelFunctionAutomat
...
+
+        var promptFunction = KernelFunctionYaml.FromPromptYaml(promptTemplate);
+
+        string result = "";
+
+        // Act
+        await foreach (string c in promptFunction.InvokeStreamingAsync<string>(this._kernel))
+        {
+            result += c;
+        }
```
Comment (RogerBarreto) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAIChatCompletion_AutoFunctionChoiceBehaviorTests.cs:200]: Avoiding abbreviations.

```suggestion
        await foreach (string content in promptFunction.InvokeStreamingAsync<string>(this._kernel))
        {
            result += content;
        }
```

**Review by RogerBarreto — APPROVED** [review]

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs (hunk 3 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:21]: The method is declared in the `IYamlTypeConverter` interface and should be implemented by the class as and instance one.

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs (hunk 4 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:27]: Nope, unfortunately the `IYamlTypeConverter` interface does not have async methods:
image

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs (hunk 7 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs:63]: I'm not aware of any plans to do so. We can always do it when/if needed.

File: dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAIChatCompletion_NoneFunctionChoiceBehaviorTests.cs (hunk 9 above)
Comment (SergeyMenshykh) [reply dotnet/src/IntegrationTests/Connectors/OpenAI/OpenAIChatCompletion_NoneFunctionChoiceBehaviorTests.cs]: Thanks, fixed.

File: dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs (hunk 6 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs]: Kept them it because it's needed and fixed the number.

Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.Yaml/PromptExecutionSettingsTypeConverter.cs]: Thanks, fixed.

File: dotnet/src/Connectors/Connectors.OpenAI.UnitTests/Settings/OpenAIPromptExecutionSettingsTests.cs (hunk 1 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.OpenAI.UnitTests/Settings/OpenAIPromptExecutionSettingsTests.cs]: Fixed.

File: dotnet/src/Functions/Functions.UnitTests/Yaml/Functions/KernelFunctionYamlTests.cs (hunk 2 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.UnitTests/Yaml/Functions/KernelFunctionYamlTests.cs]: Refactored.

**Review by RogerBarreto — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]

**Comment by SergeyMenshykh** [comment]
Linking the related issue: https://github.com/microsoft/semantic-kernel/issues/5954

**Pull request merged**